### PR TITLE
17928 added generic agm extension schema for further FE development

### PR DESCRIPTION
**Pull request by PaulGarewal**
*Issue #:* /bcgov/entity#17928

*Description of changes:*
 - added basic agm extension schema data. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-schemas license (Apache 2.0).

### CODE REVIEW

File: src/registry_schemas/schemas/agm_extension.json
Comment (severinbeauvais) [inline src/registry_schemas/schemas/agm_extension.json:18]: Would it make sense to change this to "integer"?

**Comment by severinbeauvais** [comment]
That was quick! Looks good, too.

Have a look at the build error, and wait for another review before merging (I don't feel that I'm an acceptable gatekeeper for this code).

File: src/registry_schemas/schemas/agm_extension.json
Comment (thorwolpert) [reply src/registry_schemas/schemas/agm_extension.json]: if you do a number type, you'll need a min/max, or stay with string and provide a regex (4 chars only and appropriate ranges)

File: src/registry_schemas/schemas/agm_extension.json
Comment (thorwolpert) [inline src/registry_schemas/schemas/agm_extension.json:14]: should this be a date?
How does specifying a year (containing 12 months) have a condition of no more than 6 months. That doesn't align in my head.

**Review by thorwolpert — COMMENTED** [review]
Do we have a full spec?
If not, we should get a copy of the paper version and go from there.

File: src/registry_schemas/schemas/agm_extension.json
Comment (severinbeauvais) [reply src/registry_schemas/schemas/agm_extension.json]: Is there an appropriate range for AGM Year?

File: src/registry_schemas/schemas/agm_extension.json
Comment (severinbeauvais) [inline src/registry_schemas/schemas/agm_extension.json:23]: This is also a required property ;)

**Comment by severinbeauvais** [comment]
> Do we have a full spec? If not, we should get a copy of the paper version and go from there.

We're working on it (with Mihai and Yui). It's not the same as the paper version -- it's being simplified.

File: src/registry_schemas/schemas/agm_extension.json
Comment (severinbeauvais) [reply src/registry_schemas/schemas/agm_extension.json]: This is the year of the AGM they are requesting an extension for https://preview.uxpin.com/c32b7a6fabe9dea89e09a5ae63323f485a02396c#/pages/164523638/simulate/no-panels

It's not a date. It's just a year.

File: src/registry_schemas/schemas/agm_extension.json
Comment (thorwolpert) [reply src/registry_schemas/schemas/agm_extension.json]: well, shouldn't be an unbounded integer eg 26,000 is a valid year, but just one I'm unlikely to see.

Comment (thorwolpert) [reply src/registry_schemas/schemas/agm_extension.json]: also sounds like it will require a coded validator (if 6 months is the maximum extension, it should probably be the date is was extended to)

Comment (severinbeauvais) [reply src/registry_schemas/schemas/agm_extension.json]: Yes, there will be a data field that needs to be validated down to the month, but it's not AGM Year.

Hmm, maybe we should keep this as a string, same as AGM Location Change, or change that one to an integer as well? Check with whoever is working with that one.https://github.com/bcgov/business-schemas/blob/3b55629ff95f3a2e857ca0774f14b46fcceec82c/src/registry_schemas/schemas/agm_location_change.json#L18

File: tests/unit/test_agm_extension.py
Comment (severinbeauvais) [inline tests/unit/test_agm_extension.py:48]: Also add a test for isFirstAgm.

File: src/registry_schemas/schemas/agm_extension.json
Comment (severinbeauvais) [reply src/registry_schemas/schemas/agm_extension.json]: > also sounds like it will require a coded validator

Yes, that will be done in [this ticket](https://app.zenhub.com/workspaces/entities-team-space-6143567664fb320019b81f39/issues/gh/bcgov/entity/18053).

Comment (severinbeauvais) [reply src/registry_schemas/schemas/agm_extension.json:19]: OK, so it stays as a string 👍

File: tests/unit/test_agm_extension.py
Comment (severinbeauvais) [reply tests/unit/test_agm_extension.py:48]: Thanks.

**Review by severinbeauvais — APPROVED** [review]
LGTM

File: src/registry_schemas/schemas/agm_extension.json
Comment (thorwolpert) [reply src/registry_schemas/schemas/agm_extension.json]: Sounds incorrect, based on the business requirements below, sounds like dates need to be captured.
Four requirements for AGM extension:

Rules:
An association may apply for an extension in writing giving a brief explanation why the meeting must be delayed, when the meeting ought to have been held (within 4 months of the financial year end), and the date when the meeting will be held.
Have we already granted an extension for this AGM? The client may apply for a maximum six month extension, and not beyond March 31 of the year following the original AGM date. (They may apply twice, but the total extension cannot exceed six months.)
Is the coop in Good Standing? - see details below.
Check the prior years filed Annual Report to see if it had a reported meeting date on it.  If it was filed as "no meeting held" we will not provide an extension for the current year.
Has the request for the extension come in before the date the coop ought to have held the current year's AGM by? They must make the extension request before the day they ought to have held their AGM. A coop must hold a general meeting within 4 months after the end of its financial year.  If the request is coming later than 4 months after the financial year end it will be rejected.

Comment (severinbeauvais) [reply src/registry_schemas/schemas/agm_extension.json]: There will be more data captured but we haven't agreed on names yet.

THIS is an initial schema for the AGM extension filing. There is a part 2 ticket for completing the schema later.

In THIS schema, we do want to capture the year of the AGM that is being extended. That's this field.

File: src/registry_schemas/schemas/agm_extension.json
Comment (thorwolpert) [inline src/registry_schemas/schemas/agm_extension.json:15]: Is this required?
Is this something the user is sending us, or we're calculating and sending back?
would it be more helpful to have the date of the last agm? then the other business rules can be applied and a Null/"" value could be used to infer no prior agm?

**Review by thorwolpert — COMMENTED** [review]
does the flag provide value, and is it required or calculated by us?

File: src/registry_schemas/schemas/agm_extension.json
Comment (severinbeauvais) [reply src/registry_schemas/schemas/agm_extension.json:15]: This is a required input from the user.

We do not necessarily have the date of the last AGM.

I figured we'd store this in the JSON so that the Legal API can validate the filing data (eg, if this is the first AGM then follow-on data is optional).

File: src/registry_schemas/schemas/agm_extension.json
Comment (severinbeauvais) [inline src/registry_schemas/schemas/agm_extension.json:1]: I wasn't thinking that you'd do this right away but I think it's helpful nonetheless :)

File: src/registry_schemas/schemas/agm_extension.json
Comment (severinbeauvais) [inline src/registry_schemas/schemas/agm_extension.json:21]: These are not all required. Some are only required if this is not the first AGM.

I'm not sure if we should enforce conditionally required properties here or in the Legal API.

Comment (PaulGarewal) [reply src/registry_schemas/schemas/agm_extension.json:21]: Ok I will look into this 👍

File: src/registry_schemas/schemas/agm_extension.json
Comment (severinbeauvais) [inline src/registry_schemas/schemas/agm_extension.json:27]: This is now inconsistent with the year definition in the AGM Location schema. It would be best if they were both the same. (Maybe the other one should be updated.)

What does the Legal API / db expect this field to be typed as? Do we need to do any math with this field (eg, current_year - year <= 2)?

File: src/registry_schemas/schemas/agm_extension.json
Comment (severinbeauvais) [reply src/registry_schemas/schemas/agm_extension.json:21]: Don't spend more time on this because we're still waiting for answers on what we're going to do.

The AGM Extension flow (step) analysis you need to do here is the same as what is needed to design the UI itself. It would be nice to do it once (or together)... after we get the answers. Please look for another ticket in the meantime.

File: src/registry_schemas/schemas/agm_extension.json
Comment (severinbeauvais) [reply src/registry_schemas/schemas/agm_extension.json:14]: I have no further comment on this conversation. Thor, OK to resolve it?

File: src/registry_schemas/schemas/agm_extension.json
Comment (severinbeauvais) [reply src/registry_schemas/schemas/agm_extension.json]: Obsolete conversation. Resolving.

File: src/registry_schemas/schemas/agm_extension.json
Comment (severinbeauvais) [reply src/registry_schemas/schemas/agm_extension.json]: Thanks for fixing.

File: src/registry_schemas/schemas/agm_extension.json
Comment (severinbeauvais) [reply src/registry_schemas/schemas/agm_extension.json]: Only year is now required. Thor, please resolve this conversation if OK with you.

**Review by thorwolpert — APPROVED** [review]
there's 6 of these a year, we can fix stuff later

File: src/registry_schemas/example_data/schema_data.py
Comment (severinbeauvais) [reply src/registry_schemas/example_data/schema_data.py:2721]: From [UI design](https://preview.uxpin.com/c32b7a6fabe9dea89e09a5ae63323f485a02396c#/pages/164523638/simulate/no-panels):

image

**Pull request merged**
